scripts/toolchain/fbtenv.sh

```diff
@@ -1,54 +1,195 @@
-    PATH="$SCRIPT_PATH/$TOOLCHAIN_PATH/openocd/bin:$PATH";
+    echo "yes";
+    return 0;
+}
+
+fbtenv_download_toolchain_tar()
+{
+    echo "Downloading toolchain:";
+    mkdir -p "$FBT_TOOLCHAIN_PATH/toolchain" || return 1;
+    "$DOWNLOADER" $DOWNLOADER_ARGS "$FBT_TOOLCHAIN_PATH/toolchain/$TOOLCHAIN_TAR" "$TOOLCHAIN_URL" || return 1;
+    echo "done";
+    return 0;
+}
+
+fbtenv_remove_old_tooclhain()
+{
+    printf "Removing old toolchain (if exist)..";
+    rm -rf "${TOOLCHAIN_ARCH_DIR}";
+    echo "done";
+}
+
+fbtenv_show_unpack_percentage()
+{
+    LINE=0;
+    while read -r line; do
+        LINE=$(( LINE + 1 ));
+        if [ $(( LINE % 300 )) -eq 0 ]; then
+            printf "#";
+        fi
+    done
+    echo " 100.0%";
+}
+
+fbtenv_unpack_toolchain()
+{
+    echo "Unpacking toolchain:";
+    tar -xvf "$FBT_TOOLCHAIN_PATH/toolchain/$TOOLCHAIN_TAR" -C "$FBT_TOOLCHAIN_PATH/toolchain" 2>&1 | fbtenv_show_unpack_percentage;
+    mkdir -p "$FBT_TOOLCHAIN_PATH/toolchain" || return 1;
+    mv "$FBT_TOOLCHAIN_PATH/toolchain/$TOOLCHAIN_DIR" "$TOOLCHAIN_ARCH_DIR" || return 1;
+    echo "done";
+    return 0;
 }
-main;
+
+fbtenv_clearing()
+{
+    printf "Clearing..";
+    rm -rf "${FBT_TOOLCHAIN_PATH:?}/toolchain/$TOOLCHAIN_TAR";
+    echo "done";
+    return 0;
+}
+
+fbtenv_curl_wget_check()
+{
+    printf "Checking curl..";
+    if ! curl --version > /dev/null 2>&1; then
+        echo "no";
+        printf "Checking wget..";
+        if ! wget --version > /dev/null 2>&1; then
+            echo "no";
+            echo "No curl or wget found in your PATH.";
+            echo "Please provide it or download this file:";
+            echo;
+            echo "$TOOLCHAIN_URL";
+            echo;
+            echo "And place in $FBT_TOOLCHAIN_PATH/toolchain/ dir mannualy.";
+            return 1;
+        fi
+        echo "yes"
+        DOWNLOADER="wget";
+        DOWNLOADER_ARGS="--show-progress --progress=bar:force -qO";
+        return 0;
+    fi
+    echo "yes"
+    DOWNLOADER="curl";
+    DOWNLOADER_ARGS="--progress-bar -SLo";
+    return 0;
+}
+
+fbtenv_check_download_toolchain()
+{
+    if [ ! -d "$TOOLCHAIN_ARCH_DIR" ]; then
+        fbtenv_download_toolchain || return 1;
+    elif [ ! -f "$TOOLCHAIN_ARCH_DIR/VERSION" ]; then
+        fbtenv_download_toolchain || return 1;
+    elif [ "$(cat "$TOOLCHAIN_ARCH_DIR/VERSION")" -ne "$FBT_TOOLCHAIN_VERSION" ]; then
+        fbtenv_download_toolchain || return 1;
+    fi
+    return 0;
+}
+
+fbtenv_show_usage()
+{
+    echo "fbtenv.sh - sets eviroment variables for Flipper Build Tool (FBT) and downloads toolchain";
+    echo "Warning! Not recomended to run this script mannualy! Use ./fbt !";
+    echo "Usage:";
+    printf "\tFBT_TOOLCHAIN_VERSION=.. FBT_TOOLCHAIN_PATH=.. SCRIPT_PATH=.. source fbtenv.sh\n";
+    echo "Example:";
+    printf '\tFBT_TOOLCHAIN_VERSION=10 SCRIPT_PATH=$(pwd) FBT_TOOLCHAIN_PATH=/tmp/ source scripts/toolchain/fbtenv.sh\n'
```

Comment:
can we have a simpler way to run fbtenv? use defaults when sourcing from repo root?